### PR TITLE
Updates to Kernels

### PR DESCRIPTION
**Pull request by gpleiss**
- Introduce `_create_input_grid` method for kernels, which automatically
  creates the grid for pairwise computations
- Update docs for all base kernels
- Backfill unit tests for base kernels
- Refactor SM kernel
- Introduce Cosine kernel

This is a start to fixing #249. It doesn't solve the error, but the next PR/refactor will

### CODE REVIEW

File: gpytorch/kernels/rbf_kernel.py [hunk 1]
```diff
@@ -102,6 +92,9 @@ def forward_diag(self, x1, x2):
         return res
 
     def forward(self, x1, x2):
-        lengthscales = self.log_lengthscale.exp().mul(math.sqrt(2)).clamp(self.eps, 1e5)
-        diff = (x1.unsqueeze(2) - x2.unsqueeze(1)).div_(lengthscales.unsqueeze(1))
-        return diff.pow_(2).sum(-1).mul_(-1).exp_()
+        x1_, x2_ = self._create_input_grid(x1, x2)
```
Comment (jacobrgardner) [inline gpytorch/kernels/rbf_kernel.py]: Didn't we decide we were going to pass **args in to `forward` and pass `diag=True` or `False` on to create_input_grid?

File: gpytorch/kernels/spectral_mixture_kernel.py [hunk 2]
```diff
@@ -50,35 +142,58 @@ def initialize_from_data(self, train_x, train_y, **kwargs):
...
-        if not n_dims == self.n_dims:
-            raise RuntimeError("The number of dimensions doesn't match what was supplied!")
 
-        mixture_weights = self.log_mixture_weights.view(self.n_mixtures, 1, 1, 1).exp()
-        mixture_means = self.log_mixture_means.view(self.n_mixtures, 1, 1, 1, self.n_dims).exp()
-        mixture_scales = self.log_mixture_scales.view(self.n_mixtures, 1, 1, 1, self.n_dims).exp()
-        distance = (x1.unsqueeze(-2) - x2.unsqueeze(-3)).abs()  # distance = x^(i) - z^(i)
+        if not num_dims == self.ard_num_dims:
+            raise RuntimeError(
+                "The SpectralMixtureKernel expected the input to have %d dimensionality "
```
Comment (Balandat) [inline gpytorch/kernels/spectral_mixture_kernel.py]: Can we agree on the convention of `using.format()` for string formatting throughout? I find that much clearer.

Comment (gpleiss) [reply gpytorch/kernels/spectral_mixture_kernel.py]: Sure I can fix that

File: gpytorch/kernels/rbf_kernel.py (hunk 1 above)
Comment (gpleiss) [reply gpytorch/kernels/rbf_kernel.py]: This PR doesn't address the `diag=True` and `**args` stuff. That'll be the next PR (trying to keep the PRs small). This one just adds the `_create_grid` method.

**Pull request merged**